core/src/main/java/org/fenixedu/messaging/domain/Message.java

```diff
@@ -254,14 +314,75 @@ public void delete() {
 
     public Set<String> getBccs() {
         Set<String> base = recipientsToEmails(getBccSet());
-        if (getExtraBccs() != null && !getExtraBccs().isEmpty()) {
-            Set<String> extended = Sets.newHashSet(getExtraBccs().replace(',', ' ').replace(';', ' ').split("\\s+"));
-            extended.addAll(base);
-            return extended;
-        }
+        base.addAll(getExtraBccsSet());
         return base;
     }
 
+    public Set<String> getExtraBccsSet() {
+        String extraBccs = getExtraBccs();
+        if (!Strings.isNullOrEmpty(extraBccs)) {
+            return Sets.newHashSet(getExtraBccs().replace(',', ' ').replace(';', ' ').split("\\s+"));
+        } else {
+            return Sets.newHashSet();
+        }
+    }
+
+    private static Map<Locale, Set<String>> emailsByLocale(Set<PersistentGroup> groups, Predicate<String> emailValidator) {
+        Map<Locale, Set<String>> emails = new HashMap<Locale, Set<String>>();
+        Locale defLocale = I18N.getLocale();
+        for (PersistentGroup group : groups) {
+            for (User u : group.getMembers()) {
+                UserProfile profile = u.getProfile();
+                Locale l = profile.getPreferredLocale();
+                if (l == null) {
+                    l = defLocale;
+                }
+                String email = profile.getEmail();
+                if (emailValidator.test(email)) {
+                    Set<String> localeEmails = emails.get(l);
```

Comment:
For this whole block you can just do `emails.computeIfAbsent(l, k -> new HashSet<>()).add(email)`.